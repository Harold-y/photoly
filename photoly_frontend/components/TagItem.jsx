import {Checkbox***REMOVED*** from "@chakra-ui/react";

export default function TagItem({tagName***REMOVED***){
    return (
        <Checkbox variant={"ghost"***REMOVED*** w={"100%"***REMOVED*** pl={4***REMOVED***>{tagName***REMOVED***</Checkbox >
    )
***REMOVED***