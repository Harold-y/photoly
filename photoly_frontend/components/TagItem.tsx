import {Checkbox} from "@chakra-ui/react";

interface props {
    tagName: string
}

export default function TagItem({tagName}:props){
    return (
        <Checkbox variant={"ghost"} w={"100%"} pl={4}>{tagName}</Checkbox >
    )
}