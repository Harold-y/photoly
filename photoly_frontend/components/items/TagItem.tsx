import { Box, Checkbox, CheckboxGroup, Input, useDisclosure, useOutsideClick ***REMOVED*** from "@chakra-ui/react";
import React, { ChangeEvent, KeyboardEvent, useRef, useState ***REMOVED*** from "react";
import { ContextMenu ***REMOVED*** from "../ContextMenu";
import TagContextMenu from "../contextMenus/TagContextMenu";
import { useDeleteTagMutation, useRenameTagMutation ***REMOVED*** from "../../redux/api/tagSlice";
import { useAppDispatch, useAppSelector ***REMOVED*** from "../../redux/hooks";
import { addTag, removeTag ***REMOVED*** from "../../redux/states/searchFilterSlice";

interface TagItemProps {
  tagId: number;
  tagName: string;
***REMOVED***

const TagItem: React.FC<TagItemProps> = ({ tagId, tagName ***REMOVED***) => {
  const [tag, setTag] = useState(tagName);
  const inputRef = useRef<HTMLInputElement>(null);
  const checkboxRef = useRef<HTMLInputElement>(null);
  const { isOpen, onOpen, onClose ***REMOVED*** = useDisclosure();  // control edit mode

  const [deleteRequest] = useDeleteTagMutation();
  const [renameRequest] = useRenameTagMutation();
  const curFilter = useAppSelector((state) => state.searchFilter);
  const dispatch = useAppDispatch()


  const handleRename = () => {
    onOpen();
    inputRef.current?.focus();
  ***REMOVED***;

  const callThenClose = () => {
    // empty or unmodified
    if (tag.length === 0 || tag === tagName) {
      setTag(tagName);
      onClose();
      return;
***REMOVED***

    // modified
    renameRequest({ tag_id: tagId, new_name: tag ***REMOVED***);
    onClose();
  ***REMOVED***;

  const handleEnterClose = (e: KeyboardEvent<HTMLInputElement>) => {
    if (e.key === "Enter")
      callThenClose();
  ***REMOVED***;

  useOutsideClick({
    ref: inputRef,
    handler: callThenClose
  ***REMOVED***);

  const handleDelete = () => {
    deleteRequest(tagId);
  ***REMOVED***;

  const handleChecked = (e: ChangeEvent<HTMLInputElement>) => {
    if (e.currentTarget.checked){
      dispatch(addTag(tagId))
***REMOVED***
    else{
      dispatch(removeTag(tagId))
***REMOVED***
  ***REMOVED***;

  return (
    <ContextMenu<HTMLDivElement>
      renderMenu={() =>
        <TagContextMenu
          handleRename={handleRename***REMOVED***
          handleDelete={handleDelete***REMOVED***
        />***REMOVED***
    >
  ***REMOVED***ref => (
        <Box ref={ref***REMOVED*** w="100%" pl={4***REMOVED*** display="flex">
          <CheckboxGroup value={curFilter.tagIds***REMOVED***>
            <Checkbox
              ref={checkboxRef***REMOVED***
              variant="ghost"
              value={tagId***REMOVED***
              pr={2***REMOVED***
              onChange={handleChecked***REMOVED***
            >
          ***REMOVED***isOpen ? "" : tagName***REMOVED***
            </Checkbox>
        ***REMOVED***isOpen &&
              <Input
                ref={inputRef***REMOVED***
                variant="flushed"
                value={tag***REMOVED***
                autoFocus
                onKeyDown={handleEnterClose***REMOVED***
                onChange={(e) => setTag(e.currentTarget.value)***REMOVED***
              />
    ***REMOVED***
          </CheckboxGroup>
        </Box>
      )***REMOVED***
    </ContextMenu>

  );
***REMOVED***;
export default TagItem;