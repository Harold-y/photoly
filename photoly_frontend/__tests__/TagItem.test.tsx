import {describe, expect} from "@jest/globals";
import {render} from "@testing-library/react";
import TagItem from "../components/TagItem";


describe("<TagItem/> renders correctly", () => {
    test("should render tag item on the screen", () => {
        const {container, getByText} = render(<TagItem tagName={"something"}/>)
        expect(container.firstChild).toMatchSnapshot()
        const tag = getByText(/something/i)
        expect(tag).toBeTruthy()
    })
})