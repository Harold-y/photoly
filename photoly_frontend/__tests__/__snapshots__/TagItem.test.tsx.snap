// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TagItem/> renders correctly should render tag item on the screen 1`] = `
<div
  class="css-8atqhb"
>
  <label
    class="chakra-checkbox css-1wj4kef"
  >
    <input
      class="chakra-checkbox__input"
      style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
      type="checkbox"
      value=""
    />
    <span
      aria-hidden="true"
      class="chakra-checkbox__control css-dnty2r"
    />
    <span
      class="chakra-checkbox__label css-1oeb2oe"
    >
      something
    </span>
  </label>
</div>
`;
